[Core] Add S3 public bucket fallback to handle NoCredentialsError

gangsf opened the pull request:
When S3 credentials are not available, automatically fall back to using unsigned requests (UNSIGNED signature) to access public S3 buckets.

This fixes the issue where Ray runtime environments fail to download packages from public S3 buckets when AWS credentials are not configured.

The implementation:
- First attempts to create a standard S3 client with credentials
- Tests credential availability with a ListBuckets API call
- Falls back to unsigned client if NoCredentialsError is raised
- Preserves other errors for proper error handling

Fixes: botocore.exceptions.NoCredentialsError when accessing public buckets

<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Commit:
Add S3 public bucket fallback to handle NoCredentialsError

When S3 credentials are not available, automatically fall back to using
unsigned requests (UNSIGNED signature) to access public S3 buckets.

This fixes the issue where Ray runtime environments fail to download
packages from public S3 buckets when AWS credentials are not configured.

The implementation:
- First attempts to create a standard S3 client with credentials
- Tests credential availability with a ListBuckets API call
- Falls back to unsigned client if NoCredentialsError is raised
- Preserves other errors for proper error handling

Fixes: botocore.exceptions.NoCredentialsError when accessing public buckets
Signed-off-by: Gang Zhao <gang@gang-JQ62HD2C37.local>

Commit:
Add comprehensive unit tests for S3 public bucket fallback

Added three test scenarios to verify S3 client creation behavior:

1. test_s3_client_creation_with_credentials:
   - Tests normal flow when AWS credentials are available
   - Verifies signed client is used and ListBuckets call succeeds

2. test_s3_client_creation_without_credentials:
   - Tests fallback when NoCredentialsError is raised
   - Verifies unsigned client is created with UNSIGNED signature
   - Ensures proper Config setup for public bucket access

3. test_s3_client_creation_with_other_errors:
   - Tests handling of non-credential errors (network, permissions)
   - Verifies signed client is still returned for proper error propagation
   - Ensures fallback only occurs for NoCredentialsError

All tests use comprehensive mocking to avoid external dependencies
and verify the exact behavior of the credential fallback mechanism.

Signed-off-by: Gang Zhao <gang@gang-JQ62HD2C37.local>

Commit:
Improve S3 credential detection using boto3.Session().get_credentials()

Replace network-based credential detection with efficient local check:

BEFORE:
- Used private _make_api_call('ListBuckets') method (unstable API)
- Made unnecessary network API call (slow, requires permissions)
- Required s3:ListAllMyBuckets permission
- Failed for valid but restricted credentials

AFTER:
- Use boto3.Session().get_credentials() (stable public API)
- No network calls - purely local credential check
- No special permissions required
- Works correctly with restricted credentials

Benefits:
- Faster: No network latency
- More reliable: Uses stable boto3 API
- More permissive: Works with restricted AWS credentials
- Cleaner: Removed NoCredentialsError import (no longer needed)

Updated tests to match new implementation:
- Removed test for 'other errors' (no longer applicable)
- Updated mocking to use Session().get_credentials()
- Simplified test scenarios to credentials vs no-credentials

Signed-off-by: Gang Zhao <gang@gang-JQ62HD2C37.local>

Commit:
Apply formatting fixes to S3 protocol implementation

Add blank line after import statements for better code formatting
as required by scripts/format.sh

Signed-off-by: Gang Zhao <gang@gang-JQ62HD2C37.local>

Commit:
Fix import order in S3 protocol tests

- Reorder imports to be alphabetical (sys before unittest.mock)
- Fixes linting error in CI

Signed-off-by: Gang Zhao <gang@gang-JQ62HD2C37.local>

jjyao commented on python/ray/_private/runtime_env/protocol.py:
```diff
@@ -58,7 +58,21 @@ def _handle_s3_protocol(cls):
             )
 
-        transport_params = {"client": boto3.client("s3")}
+        def create_s3_client_with_fallback():
```
No need to wrap it in a function, we can just initializing a variable called `s3_client` and pass it to `transport_params`.

gangsf replied:
Remove the wrapper and use a variable instead

Commit:
Add S3 fallback for public buckets in runtime env protocol

- Use boto3.Session().get_credentials() to detect available credentials
- Fall back to unsigned S3 client when no credentials found
- Allows access to public S3 buckets without requiring AWS credentials

Signed-off-by: Gang Zhao <gang@gang-JQ62HD2C37.local>
